### PR TITLE
feat(411): [FE] Implement View Product Batch Details

### PR DESCRIPTION
**Pull request by rmonteseba**
image

New fields added + row expand

TODOs:

- Adjust correct sources (for each of the fields)
- Handle field modifications

### CODE REVIEW

File: indigo-frontend/src/app/pages/experiment/stoichiometry/batch-detail-info-panel/batch-detail-info-panel.component.ts [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+  standalone: true,
+})
+export class BatchDetailInfoPanelComponent implements OnInit {
+  private dictionaryService = inject(BuiltInDictionaryService);
+
+  sample = input.required<ReactionOutputSample>();
+  output = input.required<ReactionOutput>();
+  reaction = input.required<Reaction>();
+
+  // Dictionary data
```
Comment (rmonteseba) [inline indigo-frontend/src/app/pages/experiment/stoichiometry/batch-detail-info-panel/batch-detail-info-panel.component.ts]: There might be some improvements here in order to store results and avoid refetch on multiple rows expand, noting here for future improvements

File: indigo-frontend/src/app/pages/experiment/stoichiometry/product-batch-summary-table/product-batch-summary-table.mock.ts [hunk 2]
```diff
@@ -0,0 +1,183 @@
+// TODO remove this mock when real data is fully available from backend
```
Comment (rmonteseba) [inline indigo-frontend/src/app/pages/experiment/stoichiometry/product-batch-summary-table/product-batch-summary-table.mock.ts:1]: Keeping mocks by the moment until having 100% complete info comming from sample notebooks in the BE, will serve as fallback in the moment for debugging or polishing UI

File: indigo-frontend/src/core/services/experiment/product-image.service.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  readonly hasError = signal<boolean>(false);
+
+  // Actual endpoint: GET /api/eln/compounds/{compoundID}/picture
+  // Returns JSON with image data (likely SVG string or base64)
+  load(compoundId: string) {
+    console.log(`ProductImageService.load(compoundId: ${compoundId})`);
+    this.setLoading(true);
+    this.setError(false);
+
+    this.service
```
Comment (rmonteseba) [inline indigo-frontend/src/core/services/experiment/product-image.service.ts]: Placeholder code, image resolution to tackled in a follow up ticket preferrably

File: indigo-frontend/src/app/pages/experiment/stoichiometry/batch-detail-info-panel/batch-detail-info-panel.component.html [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+          <!-- TODO: Implement editable text area when ready -->
+          {{ sample().structureComment || '—' }}
+        </div>
+      </div>
+
+      <div class="field-group">
+        <label class="field-label">Salt Code & Name</label>
+        <div class="field-value read-only">
+          <!-- TODO: Implement editable select with salt codes when ready -->
+          {{ output().compound?.saltCode ? (output().compound.saltCode.code + ' - ' + output().compound.saltCode.name) : '—' }}
```
Comment (denis-dzenskevich) [inline indigo-frontend/src/app/pages/experiment/stoichiometry/batch-detail-info-panel/batch-detail-info-panel.component.html]: SaltCode.name already contains code, no need to concatenate

File: indigo-frontend/src/app/pages/experiment/stoichiometry/batch-detail-info-panel/batch-detail-info-panel.component.ts (hunk 1 above)
Comment (denis-dzenskevich) [reply indigo-frontend/src/app/pages/experiment/stoichiometry/batch-detail-info-panel/batch-detail-info-panel.component.ts]: I'd suggest to cache fetched items in DictionaryService, so multiple components can benefit from it

File: indigo-frontend/src/app/pages/experiment/stoichiometry/batch-detail-info-panel/batch-detail-info-panel.component.html [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+        </div>
+      </div>
+
+      <div class="field-group">
+        <label class="field-label">Source Details</label>
+        <div class="field-value read-only">
+          <!-- TODO: Implement editable select when ready -->
```
Comment (denis-dzenskevich) [inline indigo-frontend/src/app/pages/experiment/stoichiometry/batch-detail-info-panel/batch-detail-info-panel.component.html]: I have implemented eln-dictionary-select, you can reuse it here, it will do editing and formatting for you (and also eln-enum-select for enums and eln-entered-value for EnteredValue, but the latter might have issues with styling)

File: indigo-frontend/src/app/pages/experiment/stoichiometry/editable-data-table/editable-data-table.component.html [hunk 6]
```diff
@@ -9,7 +9,23 @@ <h3 class="text-sm font-semibold p-3">{{ title() }}</h3>
   }
 </div>
 <div class="w-full overflow-x-auto border border-gray-300 rounded">
-  <table mat-table [dataSource]="dataSource()" class="w-full" style="background-color: white;">
+  <table mat-table [dataSource]="dataSource()" class="w-full" style="background-color: white;" multiTemplateDataRows>
+
+    <!-- Expand Column -->
+    @if (expandableConfig()?.enabled) {
```
Comment (denis-dzenskevich) [inline indigo-frontend/src/app/pages/experiment/stoichiometry/editable-data-table/editable-data-table.component.html:15]: I have eln-expandable-table for exactly this purpose, maybe we can unify it?

File: indigo-frontend/src/app/pages/experiment/stoichiometry/product-batch-summary-table/product-batch-summary-table.mock.ts (hunk 2 above)
Comment (denis-dzenskevich) [reply indigo-frontend/src/app/pages/experiment/stoichiometry/product-batch-summary-table/product-batch-summary-table.mock.ts:1]: What do we miss from backend?

File: indigo-frontend/src/core/services/experiment/product-image.service.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
+import { inject, Injectable, signal } from '@angular/core';
+import { ApiService } from '@/core/services/api.service';
+
+@Injectable()
+export class ProductImageService {
```
Comment (denis-dzenskevich) [inline indigo-frontend/src/core/services/experiment/product-image.service.ts]: Can we reuse eln-api-image?

File: indigo-frontend/src/core/types/entities/compound.i.ts [hunk 8]
```diff
@@ -18,4 +18,31 @@ export interface CompoundRef {
   molWeight?: EnteredValue<MolWeightUnit>;
   exactMass?: number;
   calculatedBatchMF?: string;
+  // Additional fields from ng-411
+  compoundID?: string | null;
+  molFile?: string | null;
```
Comment (denis-dzenskevich) [inline indigo-frontend/src/core/types/entities/compound.i.ts]: molFile is no longer present in CompoundRef

**Review by denis-dzenskevich — CHANGES_REQUESTED** [review]

File: indigo-frontend/src/app/pages/experiment/stoichiometry/batch-detail-info-panel/batch-detail-info-panel.component.html (hunk 5 above)
Comment (rmonteseba) [reply indigo-frontend/src/app/pages/experiment/stoichiometry/batch-detail-info-panel/batch-detail-info-panel.component.html]: Just a quick thought for future tickets 🙂

I implemented part ot this functionality this from scratch, and only later realized there might already be components that cover a similar use case. It’s possible this was already shared with the team and I may have missed it, but if not, adding a small hint in the ticket when something reusable already exists could help avoid duplicate work and keep things consistent.

No issue at all on my side, just sharing the thought in case it helps going forward 🙏

Comment (rmonteseba) [reply indigo-frontend/src/app/pages/experiment/stoichiometry/batch-detail-info-panel/batch-detail-info-panel.component.html]: @denis-dzenskevich Also, I noticed that Reactants, Reagents, and Solvents are already using dictionaries, fields, and selects as well. I don’t think this is currently done through your components, but maybe we could create a follow-up ticket to explore unifying these under eln-dictionary-select, eln-enum-select, and eln-entered-value.

File: indigo-frontend/src/app/pages/experiment/stoichiometry/editable-data-table/editable-data-table.component.html (hunk 6 above)
Comment (rmonteseba) [reply indigo-frontend/src/app/pages/experiment/stoichiometry/editable-data-table/editable-data-table.component.html:15]: Great catch! While consolidating the tables is definitely a good idea for future maintainability, from my perspective tackling it in this PR would significantly expand the scope and complexity. The two components use fundamentally different APIs `eln-editable-data-table` uses input-based column configuration while `eln-expandable-table` uses content projection with directives. 

Unifying them would require refactoring every place where `eln-editable-data-table` is currently used throughout the codebase, plus handling the additional features (`MULTI_SELECT` with custom triggers, `BUTTON` type, component outlets for expandable content) that aren't yet in your component. This architectural refactor deserves its own focused PR where we can properly test all the edge cases and migration paths without mixing it with the current feature work. Let's create a separate ticket to consolidate the table patterns, that way we can give it the attention it deserves and keep this PR focused on its original scope.

What do you think?

Comment (denis-dzenskevich) [reply indigo-frontend/src/app/pages/experiment/stoichiometry/editable-data-table/editable-data-table.component.html:15]: Created #530

File: indigo-frontend/src/app/pages/experiment/stoichiometry/batch-detail-info-panel/batch-detail-info-panel.component.html (hunk 5 above)
Comment (denis-dzenskevich) [reply indigo-frontend/src/app/pages/experiment/stoichiometry/batch-detail-info-panel/batch-detail-info-panel.component.html]: Created #531 to cover data tables

> It’s possible this was already shared with the team and I may have missed it, but if not, adding a small hint in the ticket when something reusable already exists could help avoid duplicate work and keep things consistent.

Yes, this is pain. I don't know how to solve it in a best way

File: indigo-frontend/src/core/services/experiment/product-image.service.ts (hunk 3 above)
Comment (rmonteseba) [reply indigo-frontend/src/core/services/experiment/product-image.service.ts]: Edit: fixed here by using `eln-api-image`

File: indigo-frontend/src/core/services/experiment/product-image.service.ts (hunk 7 above)
Comment (rmonteseba) [reply indigo-frontend/src/core/services/experiment/product-image.service.ts]: Yes, great call!

File: indigo-frontend/src/app/pages/experiment/stoichiometry/batch-detail-info-panel/batch-detail-info-panel.component.html (hunk 5 above)
Comment (rmonteseba) [reply indigo-frontend/src/app/pages/experiment/stoichiometry/batch-detail-info-panel/batch-detail-info-panel.component.html]: Can mention it during weekly stand (or team channel), just for the record

File: indigo-frontend/src/core/types/entities/compound.i.ts (hunk 8 above)
Comment (rmonteseba) [reply indigo-frontend/src/core/types/entities/compound.i.ts]: removed!, `StoredCompoundRef` and `VirtualCompoundRef` still conserving the ref

File: indigo-frontend/src/app/pages/experiment/stoichiometry/product-batch-summary-table/product-batch-summary-table.mock.ts (hunk 2 above)
Comment (rmonteseba) [reply indigo-frontend/src/app/pages/experiment/stoichiometry/product-batch-summary-table/product-batch-summary-table.mock.ts:1]: Helped myself with copilot here to sumarize, but this is the current status:

# Batch Detail Info Panel - Field Analysis

**Date:** February 4, 2026  
**Component:** batch-detail-info-panel.component  
**Test Data Source:** projects/438e9270-9627-4748-8c3a-87b1aace02fc/notebooks/52d8e549-13b7-4c5b-bb7e-55f947571d07/experiments/c32b481b-df75-4eab-96c9-c6da29af94e3/info

---

## Quick Background

I'm documenting the field sources for the batch detail info panel component. I've compiled a table showing:
- Which fields we're displaying in the UI (25 total)
- Where we're pulling each field from in the data structure
- The actual values returned from the real experiment above
- The mocked fallback values we're currently using

---

## Current Gap

Only **6 of 25 fields** (28.6%) are populated from the backend. The remaining 20 fields are undefined or empty.

---

## Complete Field Analysis Table

| # | Field Label | Source | Current Value (Real Data) | Mocked Value |
|---|---|---|---|---|
| 1 | Notebook Batch # | `sample.nbkBatchNumber` | `"88888888-0001-001"` | `"88888888-0001-001" (P1)` |
| 2 | Calculated Batch MW | `output.compound.molWeight.value` | `60.052 (G/MOL)` | `60.05 (G/MOL)` |
| 3 | Calculated Batch MF | `output.compound.calculatedBatchMF` | `"C2 H4 O2"` | `"C2 H4 O2"` |
| 4 | Source | `sample.source` | `undefined` | `{ id: "src1", name: "Commercial" }` |
| 5 | Structure Comments | `sample.structureComment` | `undefined` | `"Confirmed by NMR"` |
| 6 | Salt Code & Name | `output.compound.saltCode.name` | `undefined` | `"Hydrochloride" (P0 only)` |
| 7 | Source Details | `sample.sourceDetails` | `undefined` | `{ id: "sd1", name: "Purchased" }` |
| 8 | Stereoisomer Code | `output.compound.stereoisomerCode` | `undefined` | `{ id: "stereo-r-001", name: "(R)" } (P0)` |
| 9 | Precursor/Reactant IDs | `reaction.precursorReactantIds[].stringForm` | `[] (empty array)` | `N/A` |
| 10 | Salt Equivalent | `output.compound.saltEQ` | `undefined` | `0.5 (P0 only)` |
| 11 | Conversational Batch Number | `sample.strCode` | `"STR-00000001-00-001"` | `"STR-00000001-00-001" (P1)` |
| 12 | Theo. Weight | `output.theoWeight` | `{ value: 38.964, unit: "G", source: -30 }` | `{ value: 38.96, unit: "G", source: 1 }` |
| 13 | Theo. Moles | `output.theoMol` | `{ value: 0.6488, unit: "MOL", source: -30 }` | `{ value: 0.6488, unit: "MOL", source: 1 }` |
| 14 | Batch Comment | `sample.batchComment` | `undefined` | `"High purity batch"` |
| 15 | Compound State | `sample.componentState` | `undefined` | `{ id: "cs1", name: "Solid" }` |
| 16 | Virtual Compound ID | `output.compound.strCode` | `undefined` | `"P1" (P1) / "P0" (P0)` |
| 17 | Compound Protection | `sample.compoundProtection` | `undefined` | `[{ id: "cp1", name: "Nitrogen atmosphere" }]` |
| 18 | Melting Point | `sample.meltingPoint` | `undefined` | `{ lower: 45, upper: 48 }` |
| 19 | Residual Solvents | `sample.residualSolvents` | `undefined` | `[{ solvent: {...}, eq: 0.5 }]` |
| 20 | Storage Instructions | `sample.storageInstructions` | `undefined` | `[{ id: "si1", name: "Store at room temperature" }]` |
| 21 | Health Hazards | `sample.healthHazards` | `[] (empty array)` | `[{ id: "h1", name: "Flammable" }, ...]` |
| 22 | Solubility in Solvents | `sample.solubilityInSolvents` | `undefined` | `[{ solvent: {...}, solubidityType: "QUALITATIVE" }]` |
| 23 | Handling Precautions | `sample.handlingPrecautions` | `undefined` | `[{ id: "hp1", name: "Use in fume hood" }]` |
| 24 | External Supplier | `sample.externalSupplier` | `undefined` | `{ supplier: {...}, registryNumber: "SA-12345" }` |
| 25 | Purity | `sample.purityCalculations` | `undefined` | `[{ type: "HPLC", operator: "GREATER_THAN", purity: 95 }]` |

---

## Field Population Summary

**Populated:** 6 fields (28.6%)
- `Notebook Batch #`, `Calculated Batch MW`, `Calculated Batch MF`, `Conversational Batch Number`, `Theo. Weight`, `Theo. Moles`

**Empty/Missing:** 20 fields (71.4%)
- `Source`, `Structure Comments`, `Salt Code & Name`, `Source Details`, `Stereoisomer Code`, `Precursor/Reactant IDs`, `Salt Equivalent`, `Batch Comment`, `Compound State`, `Virtual Compound ID`, `Compound Protection`, `Melting Point`, `Residual Solvents`, `Storage Instructions`, `Health Hazards`, `Solubility in Solvents`, `Handling Precautions`, `External Supplier`, `Purity`


---

## Three Questions for Backend Alignment

1. **Source Verification:** Are we querying the correct paths in the data structure? I want to confirm we're looking in the right place for fields like sample.source, sample.componentState, sample.healthHazards, etc. Should we follow up this in teams?

2. **Backend Logic:** Do these fields exist in the data but aren't being returned? Or is the backend not populating them yet?

3. **Action Plan:** Can we iterate—fix 1-4 categories at a time (e.g., safety data, compound details) and re-test to confirm we're lowering the missing field count? can we complete this sources and logic in a follow up ticket? or do you think we need to tackle them here? (don't want to blow up the scope of the ticket, since we have added a lot)


CC: @denis-dzenskevich & @aamialkovich

**Comment by rmonteseba** [comment]
@denis-dzenskevich & @aamialkovich can we create a ticket for handle field modifications?

File: indigo-frontend/src/app/pages/experiment/stoichiometry/product-batch-summary-table/product-batch-summary-table.mock.ts (hunk 2 above)
Comment (denis-dzenskevich) [reply indigo-frontend/src/app/pages/experiment/stoichiometry/product-batch-summary-table/product-batch-summary-table.mock.ts:1]: Table:

1. Batch No - it's last section of `sample.nbkBatchNumber` (if nbkBatchNumber is *aaaaaaaa-bbbb-ccc*, we display *ccc*)
2. Chemical Name - `output.outputName`. Should have been renamed to "Product Name" or similar, I'll ask Roman
3. Reaction Step - useful when combining batches from multiple reactions; it's the index of model.reactions, +1. Can use constant "1" for now
4. Product Type - `output.type`
5. Reg Status - `sample.registrationStatus`
6. Total Weight - `sample.actualWeight`. Should probably be renamed to "Actual Weight"
7. Total Volume - `sample.volume`
8. Total Moles - `sample.actualMol`
9. Molarity - `sample.molarity`
10. Yield - `sample.yield`
11. Purity - `sample.purity`

Details:

1. Batch N - `sample.nbkBatchNumber`
2. Structure - use eln-api-image with URL `/api/eln/compounds/{compoundID}/picture`, where compoundID is `output.compound.compoundID`
3. Calculated Batch MW - `output.compound.molWeight`
4. Calculated Batch MF - `output.compound.calculatedBatchMF`
5. Source - `sample.source`
6. Structure Comments - `sample.structureComment`
~~7. Status - _removed from KB_~~
8. Salt Code & Name - `output.compound.saltCode`; it's DictionaryItemRef, its `name` already contains code, so no need to concatenate code+name
9. Source Details - `sample.sourceDetails`
11. Stereoisomer Code - `output.compound.stereoisomerCode`
12. Precursor/Reactant IDs - `reaction.precursorReactantIds`; for all reaction batches, its value is the same, so it's stored on Reaction level
13. Salt Equivalent - `output.compound.saltEQ`
13. Conversational Batch Number - _need to consult with BA_
14. Batch Comments - `sample.batchComment`
15. Theo Weight - `output.theoWeight`
16. Theo Moles - `output.theoMol`
17. Compound State - `sample.componentState`
18. Melting Point - `sample.meltingPoint`
19. Virtual Compound ID - _need to consult with BA_
20. Storage Instructions - `sample.storageInstructions`
21. Compound Protection - `sample.compoundProtection`
22. Residual Solvents - `sample.residualSolvents`
23. Health Hazards - `sample.healthHazards`
24. Solubility in Solvents - `sample.solubilityInSolvents`
25. Handling Precautions - `sample.handlingPrecautions`
26. External Supplier - `sample.externalSupplier`
27. Purity - `sample.purity`

Please reply, what of these data is absent from our example experiment, I will add them

File: indigo-frontend/src/core/types/entities/values.i.ts [hunk 9]
```diff
@@ -45,7 +45,7 @@ export enum EnteredValueSource {
 export interface EnteredValue<U> {
   value: number;
   unit: U;
-  source: number | 'fixed' | 'default';
+  source: EnteredValueSource | number | 'fixed' | 'default'; // TODO remove fixed and default string options
```
Comment (denis-dzenskevich) [inline indigo-frontend/src/core/types/entities/values.i.ts]: Please leave as is. It's mostly used by the backend, to track when it was modified and was it calculated or entered (or fixed or default)

File: indigo-frontend/src/core/types/entities/compound.i.ts (hunk 8 above)
Comment (denis-dzenskevich) [reply indigo-frontend/src/core/types/entities/compound.i.ts]: Actually we can just use main interface CompoundRef, as we're only displaying its fields (and sometimes will be checking for CompoundRef.type) But if you keep it, they don't have molFile as well

File: indigo-frontend/src/app/pages/experiment/stoichiometry/product-batch-summary-table/product-batch-summary-table.mock.ts (hunk 2 above)
Comment (rmonteseba) [reply indigo-frontend/src/app/pages/experiment/stoichiometry/product-batch-summary-table/product-batch-summary-table.mock.ts:1]: Please add these 15 fields to the example experiment's sample/compound data:

**Sample Fields:**
- `source`
- `sourceDetails`
- `structureComment`
- `componentState`
- `compoundProtection`
- `handlingPrecautions`
- `storageInstructions`
- `batchComment`
- Virtual compund Id (TBD)
- `precursorReactantIds`
- `meltingPoint`
- `residualSolvents`
- `solubilityInSolvents`
- `externalSupplier`
- `molarity`

**Compound Fields:**
- `saltCode`
- `saltEQ`
- `stereoisomerCode`

**For the table are missing these:**

image

(ensure Actual weight, total volume, total moles, molarity and yield are present for the two samples of the example)

---

These are all the fields are currently missing from the backend.

File: indigo-frontend/src/core/types/entities/compound.i.ts (hunk 8 above)
Comment (rmonteseba) [reply indigo-frontend/src/core/types/entities/compound.i.ts]: Cool!, removed!

File: indigo-frontend/src/app/pages/experiment/stoichiometry/product-batch-summary-table/product-batch-summary-table.mock.ts (hunk 2 above)
Comment (denis-dzenskevich) [reply indigo-frontend/src/app/pages/experiment/stoichiometry/product-batch-summary-table/product-batch-summary-table.mock.ts:1]: please use this experiment: https://indigo-eln-dev.test.lifescience.opensource.epam.com/projects/438e9270-9627-4748-8c3a-87b1aace02fc/notebooks/52d8e549-13b7-4c5b-bb7e-55f947571d07/experiments/3219017e-025d-4eab-a363-99b87914ecc5/info
 
Should have everything except precursorReactantIds, for this we need to use registered samples from another experiment, so we'd better test it later when we have editing and so on

**Review by denis-dzenskevich — APPROVED** [review]
